### PR TITLE
[kv_cache] integrated vlm code for benchmark (Stacked on #3527)

### PR DESCRIPTION
**Pull request by chohk88**
# Description
**Base branch:** `kv_cache` (PR #3527 )

1. Integrated VLM benchmark framework
    * Currently supports **Eagle2**
    * Planned support: **Paligemma**, **Qwen 2.5-VL**, etc.
2. Added custom token-generation function** for multi-modal (MM) models

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: tools/llm/run_vlm.py [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+# -----------------------------------------------------------------------------#
+# Global configuration
+# -----------------------------------------------------------------------------#
+DEVICE = torch.device("cuda:0")
+
+# Register SDPA as a standalone operator.  Converter & lowering pass are defined
+sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
+import transformers.models.qwen2.modeling_qwen2 as mq  # noqa: E402
+
+mq.ALL_ATTENTION_FUNCTIONS["flash_attention_2"] = mq.ALL_ATTENTION_FUNCTIONS["sdpa"]
```
Comment (peri044) [inline tools/llm/run_vlm.py:40]: Can you try to modify the model config after it is initialized instead of modifying the entries ? 
```
model.config._attn_implementation = "sdpa"
```

File: tools/llm/run_vlm.py [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+    return model, processor, emb_layer
+
+
+def _load_model(
+    model_name: str, device: torch.device, torch_dtype: torch.dtype
+) -> Tuple[torch.nn.Module, AutoProcessor, torch.nn.Embedding]:
+    """Dispatch helper for supported VLMs."""
+    if model_name.lower() == "eagle2":
+        return _load_eagle2(device, torch_dtype)
+    msg = f"Unsupported model: {model_name}"
```
Comment (peri044) [inline tools/llm/run_vlm.py:82]: Consider modifying the message to Encountered model: {model_name} is not supported. Supported models include nvidia/Eagle2-2B, PaliGemma2 (and others if they are supported)

File: tools/llm/run_vlm.py [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    """
+    Compile Eagle2 language model with Torch-TensorRT.
+
+    The function follows the same precision-specific flag logic used in
+    *run_llm.py* for consistency.
+    """
+    lm_wrap = _LMNoCache(language_model).to(DEVICE).eval()
+    max_seq_len = input_embeds.shape[1] + args.num_tokens
+
+    S = torch.export.Dim("seq", min=1, max=max_seq_len)
```
Comment (peri044) [inline tools/llm/run_vlm.py:119]: Consider renaming `S` to seq_len

File: tools/llm/run_vlm.py [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+        enabled_precisions = {torch.float32}
+        use_fp32_acc = True
+        use_explicit_typing = True
+    elif args.precision == "BF16":
+        enabled_precisions = {torch.bfloat16}
+    else:  # FP32
+        enabled_precisions = {torch.float32}
+
+    with torch.inference_mode():
+        exported = torch.export.export(
```
Comment (peri044) [inline tools/llm/run_vlm.py:136]: minor comment: consider renaming `exported` to `exported_program` here

File: tools/llm/run_vlm.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+    Depending on the target VLM, delegates to the appropriate compile routine.
+    """
+    torch_dtype = {
+        "FP16": torch.float16,
+        "BF16": torch.bfloat16,
+    }.get(args.precision, torch.float32)
+
+    example_embeds = torch.randn(
+        1,
+        2560,
```
Comment (peri044) [inline tools/llm/run_vlm.py:175]: Consider declaring this in a map eg: image_tokens_map = {model_name: 2560}EAGLE_IMG_TOKENS = 2560.

File: tools/llm/run_vlm.py [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+        model.language_model.config.hidden_size,
+        dtype=torch_dtype,
+        device=DEVICE,
+    )
+
+    if args.model.lower() == "eagle2":
+        return _compile_eagle2_lm(model.language_model, example_embeds, args)
+
+    msg = f"Unsupported model for compilation: {args.model}"
```
Comment (peri044) [inline tools/llm/run_vlm.py:184]: Consider changing this message similar to the comment above (by including list of supported models)

File: tools/llm/run_vlm.py [hunk 7]
```diff
@@ -0,0 +1,387 @@
...
+    return trt_mod
+
+
+def compile_torchtrt(
+    model: torch.nn.Module, args: argparse.Namespace
+) -> torch.nn.Module:
+    """
+    Front-end dispatcher mirroring *run_llm.py*’s `compile_torchtrt`.
+
+    Depending on the target VLM, delegates to the appropriate compile routine.
```
Comment (peri044) [inline tools/llm/run_vlm.py:166]: Consider making this docstring more informative. It looks like this function only compiles the LLM part of the VLM. Please mention that.

File: tools/llm/run_vlm.py [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+
+
+# -----------------------------------------------------------------------------#
+# Main driver
+# -----------------------------------------------------------------------------#
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Run VLM inference (PyTorch & TensorRT back-ends)"
+    )
+    parser.add_argument("--model", default="eagle2", help="VLM model name")
```
Comment (peri044) [inline tools/llm/run_vlm.py:210]: Let the model names be exactly same as HF model names instead of using short forms. Eg: the default should be `nvidia/Eagle2-2B`

File: tools/llm/run_vlm.py [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+        "BF16": torch.bfloat16,
+    }.get(args.precision, torch.float32)
+
+    model, processor, emb_layer = _load_model(args.model, DEVICE, dtype)
+
+    # -------------------------------------------------------------------------#
+    # 2. Input construction (image + text prompt)
+    # -------------------------------------------------------------------------#
+    url = "https://cdn.pixabay.com/photo/2019/08/08/23/33/car-4393990_1280.jpg"
+    image = Image.open(requests.get(url, stream=True).raw)
```
Comment (peri044) [inline tools/llm/run_vlm.py:257]: move this to a load image function

File: tools/llm/run_vlm.py [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+
+    if args.benchmark:
+        prompt_len = args.isl - 1792 - 26
```
Comment (peri044) [inline tools/llm/run_vlm.py:260]: What is 1792 and 26 ? Please specify them as variables and add comments indicating what they are

File: tools/llm/run_vlm.py [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+            pyt_stats = record_stats(
+                "PyTorch",
+                pyt_timings,
+                args.precision,
+                batch_size=args.batch_size,
+                compile_time_s=None,
+            )
+
+    # Register static cache lowering passes if requested
+    if args.cache == "static_v1":
```
Comment (peri044) [inline tools/llm/run_vlm.py:320]: is static_v2 not working ?

**Review by peri044 — CHANGES_REQUESTED** [review]
Added initial set of review comments. I'll try to run this example and add more comments later. Where is the vision model being compiled here ?
